tests/h/services/document_test.py

```diff
@@ -0,0 +1,126 @@
+            assert other_anno.document not in docs
+
+    def test_it_returns_each_document_only_once(
+        self, svc, groups, annotations, db_session
+    ):
+        # This will make both annotations point to the same document
+        annotations[1].document = annotations[0].document
+        annotations[2].document = annotations[0].document
+        db_session.commit()
+        docs = svc.fetch_by_groupid(groupid=groups["target_group"].pubid)
+
+        # But only one instance of the document is returned
+        assert docs == [annotations[0].document]
+
+    def test_it_returns_only_documents_with_shared_annotations_when_no_user(
+        self, annotations, groups, svc, db_session
+    ):
+        annotations[1].shared = False
+        annotations[2].shared = False
+        db_session.commit()
+
+        docs = svc.fetch_by_groupid(groupid=groups["target_group"].pubid)
+
+        assert docs == [annotations[0].document]
+
+    def test_it_returns_only_documents_with_visible_annotations_for_user(
+        self, svc, annotations, groups, users, db_session, matchers
+    ):
+        # This makes an "only me" annotation, associated with our target user
+        annotations[1].shared = False
+        annotations[1].userid = users["target_user"].userid
+
+        # This makes an "only me" annotation, associated with a different user
+        annotations[2].shared = False
+        annotations[2].userid = users["other_user"].userid
+
+        db_session.commit()
+
+        docs = svc.fetch_by_groupid(
+            groupid=groups["target_group"].pubid, userid=users["target_user"].userid
+        )
+
+        # The "only me" annotation for THIS user should be returned, but not the
+        # "only me" annotation associated with another user
+        assert docs == matchers.UnorderedList(
+            [annotations[0].document, annotations[1].document]
+        )
+
+    def test_it_returns_documents_ordered_by_last_activity_desc(
+        self, svc, annotations, groups, factories, db_session, matchers
+    ):
+        # Update the document associated with ``annotations[1]``...
+        annotations[1].document.title = "Bleep bloop"
+        db_session.commit()
+
+        # this will create the newest/latest document in the DB
+        new_annotation = factories.Annotation(
+            groupid=groups["target_group"].pubid, shared=True
+        )
+
+        docs = svc.fetch_by_groupid(groupid=groups["target_group"].pubid)
+
+        # ``new_annotation`` has the most recently-updated document, followed by
+        # the updated ``annotations[1]``, then the other two annotation's
+        # documents in reverse created (i.e. updated) order
+        assert docs == [
+            new_annotation.document,
+            annotations[1].document,
+            annotations[2].document,
+            annotations[0].document,
+        ]
+
+
+@pytest.fixture
+def users(factories):
+    return {"target_user": factories.User(), "other_user": factories.User()}
```

Comment:
I would create two separate fixture functions `target_user` and `other_user`. The same for the groups below. It just means that tests can take the user(s) and group(s) they need as direct arguments and then not have to extract them from dicts. `groupid=target_group.pubid` instead of `groupid=groups["target_group"].pubid`.

(If you want both users and both groups to be created for every test you can use `autouse=True`.)